src/state.rs

```diff
@@ -250,6 +255,62 @@ impl ProxyState {
         }
         None
     }
+
+    fn load_balance<'a>(&self, src: &Workload, svc: &'a Service) -> Option<&'a Endpoint> {
+        match svc.load_balancer {
+            None => svc.endpoints.values().choose(&mut rand::thread_rng()),
+            Some(ref lb) => {
+                let ranks = svc
+                    .endpoints
+                    .iter()
+                    .filter_map(|(_, ep)| {
+                        let Some(wl) = self.workloads.find_uid(&ep.workload_uid) else {
+                            debug!("failed to fetch workload for {}", ep.workload_uid);
+                            return None;
+                        };
+                        // Load balancer will define N targets we want to match
+                        // Consider [network, region, zone]
+                        // Rank = 3 means we match all of them
+                        // Rank = 2 means network and region match
+                        // Rank = 0 means none match
+                        let mut rank = 0;
+                        for target in &lb.routing_preferences {
+                            let matches = match target {
+                                LoadBalancerScopes::Region => {
+                                    src.locality.region == wl.locality.region
+                                }
+                                LoadBalancerScopes::Zone => src.locality.zone == wl.locality.zone,
+                                LoadBalancerScopes::Subzone => {
+                                    src.locality.subzone == wl.locality.subzone
+                                }
+                                LoadBalancerScopes::Node => src.node == wl.node,
+                                LoadBalancerScopes::Cluster => src.cluster_id == wl.cluster_id,
+                                LoadBalancerScopes::Network => src.network == wl.network,
+                            };
+                            if matches {
+                                rank += 1;
+                            } else {
+                                break;
```

Comment:
We should document if we always expect the scopes to be "size-ordered"

Reply:
Its meant to be documented here:
```
   // In FAILOVER mode, endpoint selection will prefer endpoints that match all preferences, but failover to groups of endpoints
    // that match less (or, eventually, none) preferences.
    // For instance, with `[NETWORK, REGION, ZONE]`, we will send to:
    // 1. Endpoints matching `[NETWORK, REGION, ZONE]`
    // 2. Endpoints matching `[NETWORK, REGION]`
    // 3. Endpoints matching `[NETWORK]`
    // 4. Any endpoints
```

Which is also how failover works in Istio-envoy

Reply:
I was also going to ask about this wrt the `break` being ok or not. I think it relies on the routing preferences being guaranteed ordered correctly.

Reply:
My intended implementation was to make order matter here. Open to changes though

Reply:
I think that's fine as long as folks are aware that it matters.